src/gpu/tsvd/tsvd.cu

```diff
@@ -217,11 +230,42 @@ namespace tsvd
 		Matrix<T>QReversed(Q.rows(), Q.columns());
 		col_reverse_q(Q, QReversed, context);
 		calculate_u(X, QReversed, sigma, U, context);
+
+		/**
+		 * SVD sign correction (same as svd_flip() in sklearn) -> https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/extmath.py#L499
+		   Sign correction to ensure deterministic output from SVD.
+    	   Adjusts the columns of u and the rows of v such that the loadings in the
+    	   columns in u that are largest in absolute value are always positive.
+		 */
+		std::vector<int> result_array(U.columns());
+		max_index_per_column(U, result_array, context);
+		Matrix<T>Signs(1, _param.k);
+		thrust::device_vector<int> d_results = result_array;
+		auto d_U = U.data();
+		auto d_Signs = Signs.data();
+		auto counting = thrust::make_counting_iterator <int>(0);
+		auto ptr = d_results.data();
+		thrust::for_each(counting, counting+Signs.size(), [=]__device__(int idx){
+			int u_idx = ptr[idx];
+			T val = 1.0;
+			if (d_U[u_idx] < 0.0) {
+				val = -1.0;
+			} else if (d_U[u_idx] == 0.0) {
+				val = 0.0;
+			} else {
+				val = 1.0;
+			}
+			d_Signs[idx] = val;
```

Comment:
This whole if/else can be rewritten as `d_Signs[idx] = (T(0) < d_U[u_idx]) - (d_U[u_idx] < T(0));`.